### PR TITLE
OpenCL and GPU builds

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -10,8 +11,16 @@ source:
   sha256: 9e8a7640911481134e60987d5d1e1cd157f430c3b4b38de8d36fc55c302bc299
 
 build:
-  number: 1
-  script: {{ PYTHON }} -m pip install .
+  number: {{ build_num }}
+  string: {{ impl }}_py{{ PY_VER }}h{{ PKG_HASH }}_{{ build_num }}
```
Comment (xhochy) [inline recipe/meta.yaml:15]: Shouldn't this also work? `conda-build` normally evaluates it after reading `number`
```suggestion
  string: {{ impl }}_py{{ PY_VER }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -10,8 +10,16 @@ source:
   sha256: 9e8a7640911481134e60987d5d1e1cd157f430c3b4b38de8d36fc55c302bc299
 
 build:
-  number: 1
-  script: {{ PYTHON }} -m pip install .
+  number: 2
+  string: {{ impl }}_py{{ PY_VER }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}
+  script:
+    - set "CMAKE_ARGS=%CMAKE_ARGS% -DUSE_CUDA=1"  # [impl == "cuda" and win]
+    - set "CMAKE_ARGS=%CMAKE_ARGS% -DUSE_GPU=1"   # [impl == "cpu" and win]
```
Comment (isuruf) [inline recipe/meta.yaml:17]: Is this okay? I've changed the default CPU builds to include OpenCL as well since it's a lightweight dependency.

Comment (xhochy) [reply recipe/meta.yaml:17]: Fine with me 👍

**Review by jameslamb — APPROVED** [review]
This looks great, thank you for all the hard work!

Just commenting to confirm the things I noticed:

* it's totally fine to skip the CUDA build on Windows... LightGBM does that too: https://github.com/microsoft/LightGBM/issues/6417#issuecomment-2058075202
* the CMake patch makes sense to me... I understand why you wouldn't want to statically link Boost for a conda package

**Comment by isuruf** [comment]
Thanks for the review. Merge?

**Review by xhochy — APPROVED** [review]

**Comment by timkpaine** [comment]
thanks @xhochy!

**Pull request merged**